include/realizations/catchment/Formulation_Manager.hpp

```diff
@@ -463,8 +463,7 @@ namespace realization {
                 if (directory != nullptr) {
```

Comment:
How about something like this?
I just copied the code that stats the file after the regex match.
I tested it just using std and it matched the performance of adding {{id}} to the path variable, 27s serial 3.2s parallel 56x

```suggestion
                if (directory != nullptr) {
                    // check if directory + file_pattern is a file before attempting to iterate
                    struct stat st;
                    if (stat((path + filepattern).c_str(), &st) == 0) {
                        if (S_ISREG(st.st_mode)) {
                            return forcing_params(
                                path + filepattern,
                                provider,
                                simulation_time_config.start_time,
                                simulation_time_config.end_time
                            );
                        }
                    }
```

Reply:
correction*  3.7s in 56x parallel. So not as fast as changing `path` to accept {{id}}, but probably only because I didn't stat the file when I did that, I just accepted the path without verifying if it existed. the speed difference serially is ~27.2 vs ~27.8 but my testing it just manually running it 5 times so I haven't been including the decimal place in the longer timings

Reply:
I just noticed an apparent bug in that code. We probably shouldn't be checking `S_ISREG`, since that would (I think) exclude symbolic links, which should be legitimate, as long as they point at an existent file.

All of this code kinda suffers from the syndrome of "asking permission, rather than forgiveness" - we should ideally just be trying to directly open the files in question.

Reply:
I think I misunderstood the original code a bit too, it's looping over all the files and checking matches, then returning at the first match it finds right? I mistook it to return all matching files. If it's only returning the first match I can have a go at reworking it to

1. try an open the file without stat
2. if that fails, build a list of every file in the directory
3. run regex against that list
4. if more than one match is found try to open the first match

The line between performance optimization and hack is blurry but I was wondering if it would be faster to get a string of all filenames, separated by some illegal filename character like hash or pipe, then run regex once against that string to pull out the matching files. Rather than running regex once per file?

Reply:
hopefully this is more clear than my fumbling with suggested changes https://github.com/NOAA-OWP/ngen/compare/master...JoshCu:ngen:open_files_first_ask_questions_later
How about something like this? I modified it to just try and open the file immediately, then if that fails fall back to the regex matching, which also just tries to open the file rather than using stat. The bit at the top of the diff is just inverting the `if directory != nullptr` logic to unindent the code one layer.